helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -1464,4 +1471,48 @@ private RealmAwareZkClient resolveZkClient(HelixZkClientFactory zkClientFactory,
   private String buildShardingKey() {
     return _clusterName.charAt(0) == '/' ? _clusterName : "/" + _clusterName;
   }
+
+  /**
+   * Check that not both zkAddress and ZkConnectionConfig are set.
+   * If zkAddress is not given and ZkConnectionConfig is given, check that ZkConnectionConfig has
+   * a ZK path sharding key set because HelixManager must work on single-realm mode.
+   * @param zkAddress
+   * @param helixManagerProperty
+   */
+  private void validateZkConnectionSettings(String zkAddress,
+      HelixManagerProperty helixManagerProperty) {
+    if (helixManagerProperty != null && helixManagerProperty.getZkConnectionConfig() != null) {
+      if (zkAddress != null && !zkAddress.isEmpty()) {
+        throw new HelixException(
+            "ZKHelixManager: cannot have both ZkAddress and ZkConnectionConfig set!");
+      }
+      RealmAwareZkClient.RealmAwareZkConnectionConfig connectionConfig =
+          helixManagerProperty.getZkConnectionConfig();
+      if (connectionConfig.getZkRealmShardingKey() == null || connectionConfig
+          .getZkRealmShardingKey().isEmpty()) {
+        throw new HelixException(
+            "ZKHelixManager::ZK path sharding key must be set for ZKHelixManager! ZKHelixManager "
+                + "is only available on single-realm mode.");
+      }
+      _realmAwareZkConnectionConfig = connectionConfig;
+    }
+  }
+
+  /**
+   * Resolve ZK connection info for logging purposes.
+   * @return
+   */
+  private String getZkConnectionInfo() {
+    String zkConnectionInfo;
+    if (_zkAddress == null) {
+      if (_helixManagerProperty != null && _helixManagerProperty.getZkConnectionConfig() != null) {
+        zkConnectionInfo = _helixManagerProperty.getZkConnectionConfig().toString();
+      } else {
+        zkConnectionInfo = "ZkAddr and ZkConnectionConfig are null!";
```

Comment:
nit, I think just return "None" is enough. Since otherwise, the log string will look strange.